targets/TARGET_Analog_Devices/TARGET_ADUCM4X50/TARGET_ADUCM4050/api/serial_api.c

```diff
@@ -0,0 +1,314 @@
+    obj->index = pinmap_merge(uart_tx, uart_rx);
+    MBED_ASSERT((int)obj->index != NC);
+    adi_uart_Open(obj->index,ADI_UART_DIR_BIDIRECTION,UartDeviceMem[obj->index],ADI_UART_MEMORY_SIZE,&hDevice[obj->index]);
+    serial_baud(obj, 9600);
+    serial_format(obj, 8, ParityNone, 1);
+    pinmap_pinout(tx, PinMap_UART_TX);
+    pinmap_pinout(rx, PinMap_UART_RX);
+    if (tx != NC) {
+        pin_mode(tx, PullUp);
+    }
+    if (rx != NC) {
+        pin_mode(rx, PullUp);
+    }
+    if (obj->index == STDIO_UART) {
+        stdio_uart_inited = 1;
+        memcpy(&stdio_uart, obj, sizeof(serial_t));
+    }
+}
+
+int serial_readable(serial_t *obj)
+{
+    bool bAvailable = false;
+    adi_uart_IsRxBufferAvailable(hDevice[obj->index], &bAvailable);
+    return bAvailable;
+}
+
+int serial_getc(serial_t *obj)
+{
+    int c;
+    void *pBuff;
+    uint32_t pHwError;
+    adi_uart_SubmitRxBuffer(hDevice[obj->index], &rxbuffer[obj->index], 1, 1);
+    adi_uart_GetRxBuffer(hDevice[obj->index], &pBuff, &pHwError);
+    c = (unsigned) rxbuffer[obj->index];
+    return (c);
+}
+
+int serial_writable(serial_t *obj)
+{
+    bool bAvailable = false;
+    adi_uart_IsTxBufferAvailable(hDevice[obj->index], &bAvailable);
+    return bAvailable;
+}
+
+void serial_putc(serial_t *obj, int c)
+{
+    void *pBuff;
+    uint32_t pHwError;
+    txbuffer[obj->index]= (char) c;
+    adi_uart_SubmitTxBuffer(hDevice[obj->index],&txbuffer[obj->index], 1, 1);
+    adi_uart_GetTxBuffer(hDevice[obj->index], &pBuff, &pHwError);
+    return;
+}
+
+#else
+
+void serial_init(serial_t *obj, PinName tx, PinName rx)
+{
+    uint32_t uart_tx = pinmap_peripheral(tx, PinMap_UART_TX);
+    uint32_t uart_rx = pinmap_peripheral(rx, PinMap_UART_RX);
+
+    obj->index = pinmap_merge(uart_tx, uart_rx);
+    MBED_ASSERT((int)obj->index != NC);
+
+    adi_uart_Open(obj->index, ADI_UART_DIR_BIDIRECTION, UartDeviceMem[obj->index], ADI_UART_MEMORY_SIZE, &hDevice[obj->index]);
+
+    serial_baud(obj, 9600);
+    serial_format(obj, 8, ParityNone, 1);
+
+    pinmap_pinout(tx, PinMap_UART_TX);
+    pinmap_pinout(rx, PinMap_UART_RX);
+
+    if (tx != NC) {
+        pin_mode(tx, PullUp);
+    }
+
+    if (rx != NC) {
+        pin_mode(rx, PullUp);
+    }
+
+    if (obj->index == STDIO_UART) {
+        stdio_uart_inited = 1;
+        memcpy(&stdio_uart, obj, sizeof(serial_t));
+    }
+
+    // set maximum FIFO depth
+    adi_uart_SetRxFifoTriggerLevel(hDevice[obj->index], ADI_UART_RX_FIFO_TRIG_LEVEL_14BYTE);
+
+    // enable FIFO
+    adi_uart_EnableFifo(hDevice[obj->index], true);
+}
+
+int serial_getc(serial_t *obj)
+{
+    int c;
+    uint32_t hwErr;
+
+    adi_uart_Read(hDevice[obj->index], (void *) &c, 1, false, &hwErr);
+    return (c);
+}
+
+void serial_putc(serial_t *obj, int c)
+{
+    uint32_t hwErr;
+
+    adi_uart_Write(hDevice[obj->index], &c, 1, false, &hwErr);
+    return;
+}
+
+int serial_readable(serial_t *obj)
+{
+    return 0;
```

Comment:
Is this again not supported? There is no register indicating there is something in the buffer? Same is for writeable?

Reply:
I think it's dead code, will remove.

Reply:
These functions are used, please implement them. How would I read from Serial otherwise? I believe many users use this to fetch data until it is readable/or push until its writeable?